sdk/offchain/src/identifier.rs

```diff
@@ -0,0 +1,386 @@
+// Copyright (c) The Diem Core Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::subaddress::{Subaddress, SubaddressParseError};
+use anyhow::{anyhow, Result};
+use bech32::{self, u5, FromBase32, ToBase32};
+use diem_sdk::{
+    move_types::account_address::AccountAddressParseError,
+    types::{
+        account_address::AccountAddress,
+        account_config::allowed_currency_code_string,
+        chain_id::{ChainId, NamedChain},
+    },
+};
+use std::collections::HashMap;
+use thiserror::Error;
+use url::Url;
+
+const DIEM_BECH32_VERSION: u8 = 1;
+
+/// Intent is a struct holdind data decoded from Diem Intent Identifier string
+/// https://dip.diem.com/dip-5/#format
+#[derive(Ord, PartialOrd, Eq, PartialEq, Hash, Clone)]
+pub struct Intent {
+    pub account_address: AccountAddress,
+    pub subaddress: Subaddress,
+    pub currency_code: Option<String>,
+    pub amount: Option<u64>,
+    pub hrp: String,
+}
+
+impl Intent {
+    /// Decode Diem intent identifier (https://dip.diem.com/dip-5/) int 3 parts:
+    /// 1. account identifier: account address & sub-address
+    /// 2. currency code
+    /// 3. amount
+    ///
+    /// ## Example
+    /// ```
+    /// use offchain::identifier::Intent;
+    /// let identifier = "diem://dm1p7ujcndcl7nudzwt8fglhx6wxn08kgs5tm6mz4us2vfufk?c=XUS&am=4500";
+    /// let intent = Intent::from(identifier, "dm").unwrap();
+    /// assert_eq!(intent.hrp, "dm");
+    /// assert_eq!(intent.amount, Some(4500));
+    /// ```
+    pub fn from(
+        encoded_intent_identifier: &str,
+        hrp: &str,
+    ) -> Result<Intent, IntentIdentifierError> {
+        let result: Url = Url::parse(encoded_intent_identifier)?;
+        if result.scheme() != "diem" {
+            return Err(IntentIdentifierError::Parse(format!(
+                "Unknown intent identifier scheme {}",
+                result.scheme()
+            )));
+        }
+
+        let (account_address, subaddress) = decode_account(result.host_str(), hrp)?;
+
+        let params: HashMap<String, String> = result.query_pairs().into_owned().collect();
+        let amount = normalize_amount(params.get("am"))?;
+        let currency_code = normalize_currency(params.get("c"))?;
+
+        Ok(Self {
+            account_address,
+            subaddress,
+            currency_code,
+            amount,
+            hrp: String::from(hrp),
+        })
+    }
+
+    /// Encode Intent as a Diem intent identifier (https://dip.diem.com/dip-5/).
+    ///
+    /// ## Example
+    /// ```
+    /// use offchain::identifier::Intent;
+    /// let identifier = "diem://dm1p7ujcndcl7nudzwt8fglhx6wxn08kgs5tm6mz4us2vfufk?c=XUS&am=4500";
+    /// let intent = Intent::from(identifier, "dm").unwrap();
+    /// assert_eq!(intent.to_encoded_string().unwrap(), identifier);
+    /// ```
+    pub fn to_encoded_string(&self) -> Result<String, IntentIdentifierError> {
```

Comment:
Given the only real error that can occur is if the hrp is wrong, maybe we can panic on error and convert this to a Display impl since the only errors would be programmer error

Reply:
As we talked about offline, Display leans towards human readable content which is different to the encoded string. Leaving this as is but let's revisit.